docs/sources/flow/concepts/modules.md

```diff
@@ -0,0 +1,145 @@
+
+* Retreiving the module source to run.
+* Creating a [Component controller][] for the module to run in.
+* Passing arguments to the loaded module.
+* Exposing exports from the loaded module.
+
+Module loaders typically are called `module.LOADER_NAME`. The list of module
+loader components can be found in the list of Grafana Agent Flow
+[Components][].
+
+Some module loaders may not support running modules with arguments or exports;
+refer to the documentation for the module loader you are using for more
+information.
+
+[Component controller]: {{< relref "./component_controller.md" >}}
+[Component]: {{< relref "../reference/components/" >}}
+
+## Module sources
+
+Modules are designed to be flexible, and can have their configuration retrieved
+from anywhere, such as:
+
+* The local filesystem
+* An S3 bucket
+* An HTTP endpoint
+
+Each module loader component will support different ways of retrieving module
+sources. The most generic module loader component, `module.string`, can load
+modules from the export of another Flow component:
+
+```river
+local.file "my_module" {
+  filename = "PATH_TO_MODULE"
+}
+
+module.string "my_module" {
+  source = local.file.my_module.content
+
+  arguments = {
+    MODULE_ARGUMENT_NAME_1 = MODULE_ARGUMENT_VALUE_1,
+    MODULE_ARGUMENT_NAME_2 = MODULE_ARGUMENT_VALUE_2,
+    // ...
+  }
+}
+```
+
+## Example module
+
+This example module manages a pipeline which filters out debug- and info-level
+log lines which are given to it:
+
+```river
+// argument.write_to is an optional argument which specifies where filtered
+// log lines should be sent.
+//
+// The value of the argument can be retrieved in this file with
+// argument.write_to.value.
+argument "write_to" {
+  optional = false
+}
+
+// loki.process.filter is our component which executes the filtering, passing
+// filtered logs to argument.write_to.value.
+loki.process "filter" {
+  // Drop all debug- and info-level logs.
+  stage.match {
+    selector = "{job!=\"\"} |~ \"level=(debug|info)\""
+    action   = "drop"
+  }
+
+  // Send processed logs to our argument.
+  forward_to = argument.write_to.value
+}
+
+// export.filter_input exports a value to the consumer of the module.
+export "filter_input" {
+  // Expose the receiver of loki.process so the module consumer can send
+  // logs to our loki.process component.
+  value = loki.process.filter.receiver
+}
+```
+
+It can then be used as a processing step before writing logs to Loki:
+
+```river
+loki.source.file "self" {
+  targets = LOG_TARGETS
+
+  // Forward collected logs to the input of our filter.
+  forward_to = [module.string.filter.values.logs_input]
+}
+
+local.file "log_filter" {
+  filename = "/path/to/modules/log_filter.river"
+}
+
+module.string "log_filter" {
+  source = local.file.log_filter.content
+
+  arguments = {
+    // Configure the filter to forward filtered logs to loki.echo below.
+    logs_output = [loki.write.default.receiver],
+  }
+}
+
+loki.write "default" {
```

Comment:
maybe reorder the components here in order of references? I'd expect the loki.write up top, then the file, the module, and the source.

Reply:
I'm going to leave this as-is mainly because it's a stylistic choice and we don't recommend a specific way yet; I normally write components in terms of the data flow top-down, with some exceptions like putting local.files before their usage.